src/main/scala/cognite/spark/v1/AssetHierarchyBuilder.scala

```diff
@@ -118,112 +196,173 @@ class AssetHierarchyBuilder(config: RelationConfig)(val sqlContext: SQLContext)
-      .filter(AssetsFilter(rootIds = Some(Seq(sourceRootExternalId))))
-      .compile
-      .toList
+  def fetchCdfRoot(sourceRootExternalId: String): IO[Option[Asset]] =
+    client.assets
+      .retrieveByExternalId(sourceRootExternalId)
+      .map(Some(_).asInstanceOf[Option[Asset]])
+      .handleError {
+        case e: CdpApiException if e.code == 400 && e.missing.isDefined =>
+          None.asInstanceOf[Option[Asset]]
+      }
 
-    for {
-      tree <- cdfTree
-      root = tree.find(_.externalId.contains(sourceRootExternalId.externalId))
-    } yield (root, tree)
-  }
+  def fetchCdfSubtree(root: AssetsIngestSchema): IO[List[Asset]] =
+    if (root.parentExternalId.isEmpty) {
+      // proper root
+      client.assets
+        .filter(AssetsFilter(rootIds = Some(Seq(CogniteExternalId(root.externalId)))))
+        .compile
+        .toList
+    } else {
+      // This API will error out if the subtree contains more than 100,000 items
+      // If that's a problem, we'll have to workaround by downloading it's entire hierarchy
+      // (or convince the API team to lift that limitation)
+      client.assets
+        .filter(AssetsFilter(assetSubtreeIds = Some(Seq(CogniteExternalId(root.externalId)))))
+        .compile
+        .toList
+    }
 
   def upsertRoot(
       root: AssetsIngestSchema,
       sourceRoot: Option[Asset]
-  ): IO[Unit] =
+  ): IO[Unit] = {
+    val parentId = if (root.parentExternalId.isEmpty) {
+      None
+    } else {
+      Some(root.parentExternalId)
+    }
+
+    // TODO: don't do the update when nothing is changed (waiting for Scala SDK parentExternalId)
     sourceRoot match {
       case None =>
         client.assets
-          .create(Seq(AssetsIngestSchema.toAssetCreate(root).copy(parentExternalId = None)))
-          .flatTap(x => incMetrics(itemsCreated, x.size)) *> IO.unit
-      case Some(asset) =>
-        if (isMostlyEqual(root, asset)) {
-          IO.unit
-        } else {
-          client.assets
-            .updateByExternalId(
-              Seq(
-                root.externalId -> AssetsIngestSchema
-                  .toAssetUpdate(root)
-                  .copy(parentExternalId = None)).toMap)
-            .flatTap(x => incMetrics(itemsUpdated, x.size)) *> IO.unit
+          .create(Seq(AssetsIngestSchema.toAssetCreate(root).copy(parentExternalId = parentId)))
+          .flatTap(x => incMetrics(itemsCreated, x.size))
+          .map(x => x.head)
+      case Some(asset) => {
+        val parentIdUpdate = (parentId, asset.parentId) match {
+          case (None, Some(oldParent)) =>
+            throw CdfDoesNotSupportException(
+              s"Can not make root from asset '${asset.externalId.get}' which is already inside asset '${oldParent}."
+              // TODO: include parentExternalId in the message
+                + " You might need to remove the asset and insert it again.")
+          case (None, None) => None
+          case (Some(newParent), _) => Some(SetValue(newParent))
```

Comment:
`if (newParent == asset.parentExternalId)`..

Reply:
You mean I should only do the update when it actually changes? That makes sense, however we don't do that for all the other fields, so we should do there too?

Reply:
That is the overall intention - assets that exist, but haven't changed, don't need to be updated. Maybe you can reuse `isMostlyEqual` here, or have it include `parentExternalId`

Reply:
> Maybe you can reuse isMostlyEqual here

Thats already the handled in the `case` above. I was just waiting for the `parentExternalId` and then forgot to push yesterday :/